File: src/utils/api/system-manager.ts
/**
 * 系统模块统一 API 入口
 *
 * 后端 `system_manager` IPC 命令通过 `action` 字段分发到不同子模块
 * （参照 `meta_manager` / `image_cache_manager` / `config_manager` 模式）。
 *
 * 字段名约定：后端 Params 结构体使用 `#[serde(rename_all = "camelCase")]`，
 * 故前端 params 对象的字段名一律使用 camelCase。
 *
 * 注册的 action（29 个）：
 * - game_dir（7 个）：`open_game_dir` / `open_path` / `reveal_in_explorer`
 *   / `get_game_dir` / `write_text_file` / `get_system_memory` / `set_game_dir`
 * - config（2 个）：`get_config_path` / `save_config_to_file`
 * - developer（6 个）：`is_developer_unlocked` / `unlock_developer_mode`
 *   / `lock_developer_mode` / `get_storage_dirs` / `get_system_info` / `get_cache_stats`
 * - devtools（3 个）：`open_devtools` / `close_devtools` / `is_devtools_open`
 *   （开发者模式解锁且开启时可调出 WebView2 DevTools）
 * - about（1 个）：`get_about_data`
 * - logger（3 个）：`get_log_path` / `list_log_files` / `read_log_file`
 * - http_log（2 个）：`read_http_logs` / `list_http_log_files`
 * - updater（2 个）：`check_update` / `download_and_install_update`
 * - certs（3 个）：`list_custom_certs` / `add_custom_cert` / `remove_custom_cert`
 * - ws（1 个）：`get_ws_port`（下载进度推送 WS 端口）
 */

import { invoke } from '@tauri-apps/api/core'

/**
 * 调用 system_manager IPC
 * @param action 操作名称（取自 SYSTEM_ACTIONS 常量）
 * @param params 参数对象（字段名使用 camelCase）
 */
export async function systemManager<T = unknown>(
  action: string,
  params?: unknown,
): Promise<T> {
  return invoke<T>('system_manager', { req: { action, params: params ?? null } })
}

/**
 * 所有可用的 action 名称
 *
 * 与后端 `utils::system_manager::DISPATCHER` 注册的 action 一一对应。
 * 业务代码应优先使用此常量而非裸字符串，避免拼写错误。
 */
export const SYSTEM_ACTIONS = {
  // game_dir（7 个）
  OPEN_GAME_DIR: 'open_game_dir',
  OPEN_PATH: 'open_path',
  REVEAL_IN_EXPLORER: 'reveal_in_explorer',
  GET_GAME_DIR: 'get_game_dir',
  WRITE_TEXT_FILE: 'write_text_file',
  GET_SYSTEM_MEMORY: 'get_system_memory',
  SET_GAME_DIR: 'set_game_dir',
  // config（2 个）
  GET_CONFIG_PATH: 'get_config_path',
  SAVE_CONFIG_TO_FILE: 'save_config_to_file',
  // developer（6 个）
  IS_DEVELOPER_UNLOCKED: 'is_developer_unlocked',
  UNLOCK_DEVELOPER_MODE: 'unlock_developer_mode',
  LOCK_DEVELOPER_MODE: 'lock_developer_mode',
  GET_STORAGE_DIRS: 'get_storage_dirs',
  GET_SYSTEM_INFO: 'get_system_info',
  GET_CACHE_STATS: 'get_cache_stats',
  // devtools（3 个）—— 调出/关闭/查询 WebView2 开发者工具
  // 后端 require_dev_mode() 校验 DeveloperUnlocked && DeveloperMode，普通用户无法触发
  OPEN_DEVTOOLS: 'open_devtools',
  CLOSE_DEVTOOLS: 'close_devtools',
  IS_DEVTOOLS_OPEN: 'is_devtools_open',
  // about（1 个）
  GET_ABOUT_DATA: 'get_about_data',
  // logger（3 个）
  GET_LOG_PATH: 'get_log_path',
  LIST_LOG_FILES: 'list_log_files',
  READ_LOG_FILE: 'read_log_file',
  // http_log（2 个）—— 联机 API 调用追踪
  READ_HTTP_LOGS: 'read_http_logs',
  LIST_HTTP_LOG_FILES: 'list_http_log_files',
  // updater（4 个）—— Windows 便携版自实现 + macOS/Linux 转发官方 plugin
  CHECK_UPDATE: 'check_update',
  DOWNLOAD_AND_INSTALL_UPDATE: 'download_and_install_update',
  // Windows 后台静默下载新版本到 appdata/last.exe + 退出时应用替换
  DOWNLOAD_UPDATE_TO_APPDATA: 'download_update_to_appdata',
  APPLY_PENDING_UPDATE: 'apply_pending_update',
  // certs（3 个）—— 自定义 TLS 证书管理
  LIST_CUSTOM_CERTS: 'list_custom_certs',
  ADD_CUSTOM_CERT: 'add_custom_cert',
  REMOVE_CUSTOM_CERT: 'remove_custom_cert',
  // deeplink（3 个）—— molaunch:// 协议注册状态查询/注册/卸载（便携版用）
  GET_DEEPLINK_STATUS: 'get_deeplink_status',
  REGISTER_DEEPLINK: 'register_deeplink',
  UNREGISTER_DEEPLINK: 'unregister_deeplink',
  // ws（1 个）—— 下载进度推送 WebSocket 端口
  GET_WS_PORT: 'get_ws_port',
} as const

/** action 名称类型 */
export type SystemAction = typeof SYSTEM_ACTIONS[keyof typeof SYSTEM_ACTIONS]

File: src-tauri/src/utils/system_manager.rs
//! 系统模块统一分发逻辑（system_manager 的工具实现）
//!
//! 使用 `utils::dispatcher::Dispatcher` 注册式分发，24 个 system action 在
//! `once_cell::sync::Lazy` 初始化时注册到 DISPATCHER。
//! 非 Result 返回的命令（`is_developer_unlocked` / `get_storage_dirs` /
//! `get_system_info` / `get_log_path` / `list_log_files` / `list_custom_certs`）
//! 需在 handler 内用 `serde_json::to_value` 包装。

use once_cell::sync::Lazy;
use serde::Deserialize;
use tauri::AppHandle;

use crate::commands::system::{
    about::get_about_data,
    config::{get_config_path, save_config_to_file},
    developer::{
        get_cache_stats, get_storage_dirs, get_system_info, is_developer_unlocked,
        unlock_developer_mode,
    },
    game_dir::{
        get_game_dir, get_system_memory, open_game_dir, open_path, reveal_in_explorer,
        set_game_dir, write_text_file,
    },
    updater,
};
use crate::handler;
use crate::logger::{get_log_path, list_log_files, read_log_file};
use crate::state::AppState;
use crate::utils::dispatcher::{ActionRequest, Dispatcher};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PathParams {
    path: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct WriteTextFileParams {
    path: String,
    content: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ReadLogFileParams {
    filename: String,
}

/// 读取 HTTP 日志的参数
///
/// - `date`: 日期字符串（`YYYY-MM-DD`），None 表示今天
/// - `limit`: 最多返回条数（从末尾截取最新的），None 表示全部
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ReadHttpLogsParams {
    date: Option<String>,
    limit: Option<usize>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SetGameDirParams {
    game_dir: String,
}

/// 添加自定义证书的参数（源文件路径）
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AddCustomCertParams {
    path: String,
}

/// 删除自定义证书的参数（证书文件名）
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RemoveCustomCertParams {
    filename: String,
}

static DISPATCHER: Lazy<Dispatcher> = Lazy::new(|| {
    let mut d = Dispatcher::new();

    d.register("open_game_dir", handler!(state, _app, _params, {
        open_game_dir(&state).await?;
        Ok(serde_json::Value::Null)
    }));
    d.register("open_path", handler!(_state, _app, params, {
        let p: PathParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        open_path(p.path).await?;
        Ok(serde_json::Value::Null)
    }));
    d.register("reveal_in_explorer", handler!(_state, _app, params, {
        let p: PathParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        reveal_in_explorer(p.path).await?;
        Ok(serde_json::Value::Null)
    }));
    d.register("get_game_dir", handler!(state, _app, _params, {
        let r = get_game_dir(&state).await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));
    d.register("write_text_file", handler!(_state, _app, params, {
        let p: WriteTextFileParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        write_text_file(p.path, p.content).await?;
        Ok(serde_json::Value::Null)
    }));
    d.register("get_system_memory", handler!(_state, _app, _params, {
        let r = get_system_memory().await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));
    d.register("set_game_dir", handler!(state, _app, params, {
        let p: SetGameDirParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        set_game_dir(&state, p.game_dir).await?;
        Ok(serde_json::Value::Null)
    }));

    d.register("get_config_path", handler!(_state, _app, _params, {
        let r = get_config_path().await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));
    d.register("save_config_to_file", handler!(state, _app, _params, {
        save_config_to_file(&state).await?;
        Ok(serde_json::Value::Null)
    }));

    // is_developer_unlocked / get_storage_dirs / get_system_info 返回非 Result，
    // handler 内用 Ok(to_value(r)?) 包装。
    d.register("is_developer_unlocked", handler!(_state, _app, _params, {
        let r = is_developer_unlocked();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    d.register("unlock_developer_mode", handler!(_state, _app, _params, {
        unlock_developer_mode()?;
        Ok(serde_json::Value::Null)
    }));
    d.register("get_storage_dirs", handler!(_state, _app, _params, {
        let r = get_storage_dirs();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    d.register("get_system_info", handler!(_state, _app, _params, {
        let r = get_system_info();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    d.register("get_cache_stats", handler!(_state, _app, _params, {
        let r = get_cache_stats().await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));

    d.register("get_about_data", handler!(_state, _app, _params, {
        let r = get_about_data().await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));

    // get_log_path / list_log_files 返回非 Result（String / Vec<String>），
    // handler 内用 Ok(to_value(r)?) 包装。read_log_file 返回 Result<String>。
    d.register("get_log_path", handler!(_state, _app, _params, {
        let r = get_log_path();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    d.register("list_log_files", handler!(_state, _app, _params, {
        let r = list_log_files();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    d.register("read_log_file", handler!(_state, _app, params, {
        let p: ReadLogFileParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        let r = read_log_file(p.filename)?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));

    d.register("read_http_logs", handler!(_state, _app, params, {
        let p: ReadHttpLogsParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        let r = crate::minecraft::online::http_log::read_http_logs(
            p.date.as_deref(),
            p.limit,
        );
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));

    d.register("list_http_log_files", handler!(_state, _app, _params, {
        let r = crate::minecraft::online::http_log::list_http_log_files();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));

    // Windows 便携版自实现 + macOS/Linux 转发官方 plugin
    d.register("check_update", handler!(state, app, _params, {
        let r = updater::check_update(&state, &app).await?;
        serde_json::to_value(r).map_err(|e| e.to_string())
    }));
    d.register("download_and_install_update", handler!(_state, app, params, {
        let p: updater::UpdateInfo = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        updater::download_and_install(&app, p).await?;
        Ok(serde_json::Value::Null)
    }));

    // Windows 便携版后台静默下载新版本到 appdata/last.exe
    d.register("download_update_to_appdata", handler!(_state, _app, params, {
        let p: updater::UpdateInfo = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        let downloaded = updater::download_update_to_appdata(p).await?;
        Ok(serde_json::to_value(downloaded).map_err(|e| e.to_string())?)
    }));

    // 退出时检查并应用待安装更新（last.exe → 替换主 exe）
    d.register("apply_pending_update", handler!(_state, app, _params, {
        let has_update = updater::apply_pending_update(&app).await?;
        Ok(serde_json::to_value(has_update).map_err(|e| e.to_string())?)
    }));

    d.register("list_custom_certs", handler!(_state, _app, _params, {
        let r = crate::certs::list_custom_certs();
        Ok(serde_json::to_value(r).map_err(|e| e.to_string())?)
    }));
    // 添加自定义证书（从源路径复制 PEM 文件到 certs 目录）
    d.register("add_custom_cert", handler!(_state, _app, params, {
        let p: AddCustomCertParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        crate::certs::add_custom_cert(&p.path)?;
        Ok(serde_json::Value::Null)
    }));
    // 删除自定义证书（按文件名删除 certs 目录下对应文件）
    d.register("remove_custom_cert", handler!(_state, _app, params, {
        let p: RemoveCustomCertParams = serde_json::from_value(params)
            .map_err(|e| format!("参数解析失败: {}", e))?;
        crate::certs::remove_custom_cert(&p.filename)?;
        Ok(serde_json::Value::Null)
    }));

    // ws（1 个）：获取 WebSocket 服务器端口 + 鉴权 token（前端建 WS 连接用）
    // 返回 {port: u16, token: string}，port=0 表示 WS 服务器尚未启动
    // token 用于客户端建连后首条消息鉴权，防止本机其他进程窃听下载进度
    d.register("get_ws_port", handler!(state, _app, _params, {
        let port = state.ws_port.get().copied().unwrap_or(0u16);
        let token = state.ws_token.get().cloned().unwrap_or_default();
        Ok(serde_json::json!({
            "port": port,
            "token": token,
        }))
    }));

    d
});

/// 分发入口
pub async fn dispatch(
    state: AppState,
    app: AppHandle,
    req: ActionRequest,
) -> Result<serde_json::Value, String> {
    DISPATCHER.dispatch(state, app, req).await
}
